### PR TITLE
Update ifutil.py - tweaking the ifutil to insert the hostname to corr…

### PR DESCRIPTION
**Pull request by LubomirH**
…ect place

This is my first pull request, so I hope I am doing it correctly.

I have fought with a TurnKey appliance on Proxmox for quite a while to make it register the hostname with my DHCP server.

I found out that the hostname was written below the hotplug interface. When I finally fixed it, I found out that after each reboot something is always removing it again.  It turned out to be the PVE integrations, to prevent it I had to create an ignore file "touch /etc/network/.pve-ignore.interfaces". It should probably be a part of the initial TurnKey config while deploying on Proxmox.

I hope I understood the loop correctly and my edits are OK, feel free to adjust it.

Iam also not sure if how I edited the loop didn't introduce any issue with some other interfaces, but I tested it thoroughly with auto eth0, then with auto lo+auto eth0 combination and finaly with auto lo+auto eth0+allow-hotplug eth1, what is I guess the default for Proxmox PVE.

It would be great if you could include this tweak to the 2.0.6 version before you push it to the APT.

thanks

Lubomir

PS: this is how looks my file now

<interfaces file>
# UNCONFIGURED INTERFACES
# remove the above line if you edit this file

auto lo
iface lo inet loopback

auto eth0
iface eth0 inet dhcp
        hostname wordpressmulti

allow-hotplug eth1
iface eth1 inet dhcp
</interfaces file>

### CODE REVIEW

**Comment by NitrogenUA** [comment]
Hi @LubomirH! It seems to me that the purpose of `_preproc_if` method is to retain **existing** interface settings as opposed to adding new ones. Therefore wouldn't it be more fitting to place the code within `set_dhcp` method? That's how it works on 2.0.5, but was since removed in 2.0.6. Perhaps @JedMeister can clarify desired approach.

**Comment by JedMeister** [comment]
Hi @LubomirH - deep apologies on such a slow response. Before I go any further, thank you for your contribution! The fact that you care enough about TurnKey to want to contribute is awesome! :smile: 

Thanks tons to you too @NitrogenUA. It's awesome that you jumped in and shared your opinion and insight. Apologies that I'm slow to follow up on your ping.

----
In response to your post @LubomirH:

> I found out that the hostname was written below the hotplug interface.

Do you mean that there was a hostname entry for eth1, but not eth0?

FWIW it's meant to be listed for both interfaces (not lo). The default interfaces file we provide (before it's touched by anything/anyone) includes a `hostname` entry for both interfaces. Assuming a hostname of 'core', it looks like this:
```
# UNCONFIGURED INTERFACES
# remove the above line if you edit this file

auto lo
iface lo inet loopback

auto eth0
iface eth0 inet dhcp
 hostname core

allow-hotplug eth1
iface eth1 inet dhcp
 hostname core
```

If the hostname was missing from eth0, then that certainly sounds like some buggy behaviour going on somewhere.

I do know that Confconsole ~v2.0.5 (the version you would have had by default) had some buggy code (under some non-standard scenarios would mangle interfaces files). Also as you also discovered, by default the file is actually managed by Proxmox (I sort of wish they'd overwrite our default so at least then Confconsole wouldn't allow you to edit it). So where the actual original bug (that stripped eth0's hostname) came from is unclear. It sounds like Proxmox may have been contributing, but perhaps Confconsole was at fault?

Either way, there is no such bug in v2.0.6. I have just re-tested now and whilst that will not add hostname if it's not already there, it will include it if it is.

I'm not sure if we want to force it to be there if it's not in the original file?

If you can be bothered and hopefully to put your mind at rest (or perhaps prove me wrong) could you check locally by launching a fresh appliance, installing Confconsole v2.0.6 immediately and double checking that the interfaces file is sane (your example that you showed is fine).

> [...] to prevent it I had to create an ignore file "touch /etc/network/.pve-ignore.interfaces". It should probably be a part of the initial TurnKey config while deploying on Proxmox

I'm open to improving the situation, but I'm not sure whether we'd want to do that by default.

I would argue that static IPs are generally the desire for servers and as such my personal workflow is to set a static IP via the Proxmox host at launch time, then I never touch the networking part of Confconsole. That relies on the Proxmox being able to write to the interfaces file. That not working OOTB would be a major PITA for me and no doubt others.

Maybe it's something that we should discuss with the Proxmox guys? Perhaps there are some changes that could be made on their side to improve the situation?

OTOH making it easier and/or more obvious for other usage scenarios sounds like a great idea!

I'm not sure of the best way to handle that though? Perhaps we should try to detect if Proxmox is managing the interfaces file managed and don't allow Confconsole to change it while it's managed by PVE? TBH I'm not 100% sure how we could detect that. (Maybe Proxmox could be encouraged to remove the header line when the file is managed by PVE?)

Although we can work out if it's LXC (`systemd-detect-virt`) and then check for that file you note (`/etc/network/.pve-ignore.interfaces`). Unless that exists, lock the interfaces from being edited by confconsole? Then we could have a setting to change that (which would write the file). Maybe that's good enough? What do you think?

----

So to summarise:

- A bug removed `hostname` from the `eth0` interface. May have been old confconsole, may have been Proxmox? Needs further investigation. Issue opened: https://github.com/turnkeylinux/tracker/issues/1832
- Current confconsole won't add hostname if it doesn't already exist, but won't remove it if it does.
    - Do we want to ensure that hostname is always added? I'd potentially be open to that change...
- When running as a PVE LXC guest, things could be improved re making it clear that the interfaces file is managed by default. I'd be interested in hearing more thoughts on that. Issue opened: https://github.com/turnkeylinux/tracker/issues/1833

**Comment by NitrogenUA** [comment]
Related [upstream issue](https://bugzilla.proxmox.com/show_bug.cgi?id=2838).

**Comment by JedMeister** [comment]
Hi @LubomirH - as per my note (see https://github.com/turnkeylinux/tracker/issues/1832#issuecomment-1610511719) I've implemented changes in confconsole that should resolve the primary underlying issue this PR was seeking to address. It doesn't resolve your other concern regarding PVE controlling the interfaces file, but I'll think about that some more and see if I can improve things as per https://github.com/turnkeylinux/tracker/issues/1833

Thanks again for your contribution, even though we didn't use. I still really appreciate you making the effort. Take care and perhaps see you about again another day... :smile:
